[jieqiboh] iP

jieqiboh opened the pull request:
### DukePro
> I'm crazy Like Mel Gibson.

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _FAST_ to use. 

All you need to do is,

1. download it from [here](https://github.com/jieqiboh/ip/releases/download/A-Jar/ip.jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's the java `main` method:


```java
public static void main(String[] args) {

    new Duke(LISTPATH).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

jieqiboh committed:
Gave chatbot a new name, added initial skeleton

jieqiboh committed:
Added echo feature

jieqiboh committed:
Added ability to mark and unmark tasks as done

jieqiboh committed:
Added Todos, Events and Deadlines

jieqiboh committed:
Added basic text-ui test

jieqiboh committed:
Added DukeExceptions which handles Duke specific errors

jieqiboh committed:
Added ability to delete tasks and used ArrayList to store tasks

jieqiboh committed:
Added Save Feature, as well as TaskListFileHandler class to handle read and writes to database file

jieqiboh committed:
Merge branch 'Level-7'

jieqiboh committed:
Added DateTime support in Deadline and Event Tasks

jieqiboh committed:
Merge branch 'branch-Level-8'

jieqiboh committed:
Updated Storage and TaskList to write to an ArrayList first

jieqiboh committed:
Added Parser and Ui implementations

jieqiboh committed:
Merge branch 'A-MoreOOP'

jieqiboh committed:
Organised classes into duke package

jieqiboh committed:
Merge branch 'A-Packages'

jieqiboh committed:
Merge branch 'add-gradle-support'

jieqiboh committed:
Added JUnit Tests

jieqiboh committed:
Merge branch 'A-JUnit'

jieqiboh committed:
Updated runtest and runtest2 files

jieqiboh committed:
Added more javadocs to public methods

jieqiboh committed:
Updated formatting to follow coding standard

jieqiboh committed:
Merge branch 'branch-A-JavaDoc'

jieqiboh committed:
Fixed merge conflict between javadoc and coding-standard

jieqiboh committed:
Added Find feature to search task descriptions

jieqiboh committed:
Merge branch 'branch-Level-9'

jieqiboh committed:
Added checkstyle, updated formatting

jieqiboh committed:
Merge branch 'branch-A-Checkstyle'

jieqiboh committed:
Changed application to use JavaFX and implemented a basic UI

jieqiboh committed:
Merge branch 'branch-Level-10'

yisiox commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,53 @@
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
```
Nit: perhaps a line break was intended here.

yisiox commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,149 @@
+import java.util.Scanner;
+
+/**
+ * Takes in and processes user input according to the available bot commands
```
Should JavaDoc sentences end with a period? I noticed the same in other places as well.

yisiox commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,149 @@
+            case "bye":
+                System.out.println("Goodbye!");
+                return false;
+                // Fallthrough
```
I like the use of fallthrough here!

yisiox commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,149 @@
+     * @return boolean that denotes whether to continue processing
+     */
+    public boolean processCmd(String command) {
+        String border = "____________________________________________________________";
```
Suggestion to make this a clean constant with explicit width:
```
private static final String BORDER = "_".repeat(50);
```

yisiox commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,149 @@
+/**
+ * Takes in and processes user input according to the available bot commands
+ */
+public class Parser {
```
Is there a reason why the `Parser` interacts with the user without the `Ui`?

yisiox commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,115 @@
+    private String filepath;
+
+    /**
+     * Carries out operations like reading and writing to a specified database file
```
Should there be a line break before the tags? I noticed the same issue in other places as well.

yisiox commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,115 @@
+
+    /**
+     * Given a list of strings, writes to a file
+     * @param stringTaskList that is a list of tasks in their string database representation
```
Perhaps `{@code stringTaskList}` can help the generated documentation look cleaner.

yisiox commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    /**
+     * Overloaded Task constructor
```
Suggested to describe overloaded constructor's function: `Class constructor specifying completion status`

yisiox commented on src/main/java/duke/Dates.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class Dates {
+    // Looks for pattern "dd/MM/yyyy HHmm", used for input purposes
+    static final Pattern DATE_INPUT_PATTERN = Pattern.compile("^(0?[1-9]|[12][0-9]|3[01])/(0?[1-9]|1[0-2])/"
```
I like the use of regex!

yisiox commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,110 @@
+        System.setOut(ps);
+        // process the user input
+        try {
+            this.parser.processCmd(input);
```
Nit: `this` is unnecessary.

yisiox commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class Deadline extends Task {
+
+    protected String by;
```
Any reason to not store `by` as a `LocalDateTime` object?

yisiox reviewed (commented):
Overall no major coding standard violations and code was easy to read. JavaDocs can be improved with line breaks, alignment and code style formatting. I feel that the `Ui` class could be used more and would improve the readability of interactions with the user.

jieqiboh committed:
Reformat JavaDocs comments

jieqiboh committed:
Add Assertions

jieqiboh committed:
Merge pull request #1 from jieqiboh/branch-A-Assertions

Add Assertions

jieqiboh committed:
Change names of constants referencing image paths

jieqiboh committed:
Merge pull request #2 from jieqiboh/branch-A-CodeQuality

Improve Code Quality

jieqiboh committed:
Undo Feature

There is no way for a user to undo an operation they performed
previously.

A user may accidentally delete a Task, and want to undo it.

Let's upgrade Steve and add a way to maintain the previous state
of the list of tasks, so that undoing is possible.

jieqiboh committed:
Merge branch 'branch-BCD-Extension'

javierng2knus commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,183 @@
+     * @param command taken in by scanner
+     * @return boolean that denotes whether to continue processing
+     */
+    public boolean processCmd(String command) {
```
Great naming of boolean members and methods. 👍 

Could the `processCmd` method possibly be named to sound more like a boolean or adjust it such that `processCmd` returns the type of command and the main `Duke` class or some other handles the execution of the commands.

javierng2knus commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,115 @@
+        // Clear and init data/duke.txt
+        fileHandler.clearFile();
+
+        List<String> input = new ArrayList<>();
```
I've noticed that the input variable could be clearer as a collection of strings by naming it to either `inputs` or `inputStrings` but other than that the rest of the naming for a collection of objects seems good. 👍

javierng2knus commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,67 @@
+     * @param todoTask is the todo Task object
+     * @return Task the string rep of Todo in the database
+     */
+    public static String todo2Db(Todo todoTask) {
```
Interesting and great way of shortening the function name while maintaining clarity!
Alone it might not be clear enough but with the comprehensive JavaDoc, it's great!👍

javierng2knus reviewed (commented):
Overall well done!
Some minor nitpicks regarding the naming convention [here](https://se-education.org/guides/conventions/java/intermediate.html), but other than that, LGTM! 👍

jieqiboh committed:
Priority Feature

There is no way for a user to assign a priority value to tasks
currently.

A user may want to track which tasks are the most urgent to address.

Let's upgrade Steve and add a way to assign an integer priority
value to tasks, so that this is possible.

jieqiboh committed:
Merge branch 'branch-priority'

jieqiboh committed:
Add user guide

jieqiboh committed:
Merge branch 'branch-A-UserGuide'

jieqiboh committed:
Fix typos in User Guide

jieqiboh committed:
Resolve styling issues

jieqiboh committed:
Update DB_PATH to not use hardcoded file path

jieqiboh committed:
Update Ui to include greeting and proper exit
